ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/API.hs

```diff
@@ -0,0 +1,239 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE RankNTypes       #-}
+module Ouroboros.Storage.ImmutableDB.API
+  ( ImmutableDB(..)
+  , withDB
+  , getCurrentEpoch
+  , Iterator(..)
+
+  , IteratorResult(..)
+  , iteratorToList
+  , blobProducer
+
+  , module Ouroboros.Storage.ImmutableDB.Types
+  ) where
+
+import Control.Monad.Catch (MonadMask, bracket)
+import Control.Monad.Except (MonadError, ExceptT(ExceptT), runExceptT,
+                             throwError)
+
+import Data.Function (on)
+import Data.ByteString (ByteString)
+import Data.ByteString.Builder (Builder)
+
+import GHC.Stack (HasCallStack)
+
+import Pipes (Producer, yield, lift)
+
+import Ouroboros.Storage.FS.Class.Types
+import Ouroboros.Storage.ImmutableDB.Types
+
+
+-- | Open the database using the given function, perform the given action
+-- using the database, and closes the database using its 'closeDB' function,
+-- in case of success or when an exception was raised.
+withDB :: (HasCallStack, MonadMask m)
+       => m (Either ImmutableDBError (ImmutableDB m))
+          -- ^ How to open the database
+       -> (ImmutableDB m -> m (Either ImmutableDBError a))
+          -- ^ Action to perform using the database
+       -> m (Either ImmutableDBError a)
+withDB openDB action = runExceptT $
+    bracket (ExceptT openDB) (ExceptT . closeDB) (ExceptT . action)
+
+-- | API for the 'ImmutableDB'.
+--
+-- See "Ouroboros.Storage.ImmutableDB.Impl" for an overview of the immutable
+-- database.
+data ImmutableDB m = ImmutableDB
+  { -- | Close the database.
+    --
+    -- Idempotent.
+    --
+    -- __Note__: Use 'withDB' instead of this function.
+    closeDB
+      :: HasCallStack => m (Either ImmutableDBError ())
+      -- TODO remove this operation from the public API, replace it with a
+      -- reopen + validate function that can be used in case of an error.
+      --
+      -- The idea is that the 'ImmutableDB' handle/record can be stored in a
+      -- Reader (instead of State), so in case of an error, we can reopen the
+      -- same database (after validation) instead of having to create a new
+      -- 'ImmutableDB' record for the same on-disk database.
+      --
+      -- We should still expose this field in an internal record so it can be
+      -- used by 'withDB'.
+
+    -- | Return 'True' when the database is open.
+  , isOpen
+      :: HasCallStack => m Bool
+
+    -- | Return the next free 'RelativeSlot' in the current epoch as an
+    -- 'EpochSlot'.
+    --
+    -- Throws a 'ClosedDBError' if the database is closed.
+  , getNextEpochSlot
```

Comment:
This one might not be very useful without some transactional guarantees. One use case: getting the latest block that was written. I'd want to get the next epoch slot, and take the previous one, then read it. If there are concurrent appends, I may get the wrong answer.

Reply:
Right. But in your use case, you only risk reading an epoch slot that is no longer the last one.

However, if you get the next epoch slot and want to append to it, it might fail when a concurrent append happens. So indeed, there needs to be some transactional guarantee.